File: SeaBattle/SeaBattle.CSharp/Board.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace SeatBattle.CSharp
{
    public class Board : Control
    {
        private const int CellSize = 25;

        private static readonly Rect BoardRegion = new Rect(0, 0, 10, 10);

        private readonly BoardCell[,] _cells;
        private readonly List<Ship> _ships;
        private DraggableShip _draggedShip;
        private readonly Random _rnd;
        private readonly bool _drawShips;

        public Board():this(true){}

        public Board(bool drawShips)
        {
            _drawShips = drawShips;
            _cells = new BoardCell[10, 10];
            _ships = new List<Ship>();
            _rnd = new Random(DateTime.Now.Millisecond);
            Mode = BoardMode.Design;
            Margin = Padding.Empty;

            CreateBoard();
        }

        /// <summary>
        ///     Creates board header cell ar a given position
        /// </summary>
        /// <param name="x">X coordnate of cell</param>
        /// <param name="y">Y coordinate of cell</param>
        /// <param name="text">Text of cell</param>
        /// <returns></returns>
        private static Label CreateHeaderCell(int x, int y, string text)
        {
            var cell = new Label
                           {
                               AutoSize = false,
                               BackColor = Color.Transparent,
                               TextAlign = ContentAlignment.MiddleCenter,
                               Text = text,
                               Location = new Point(x, y),
                               Width = CellSize,
                               Height = CellSize
                           };
            return cell;
        }

        /// <summary>
        ///     Creates board headers
        /// </summary>
        private void CreateHeaders()
        {
            for (var i = 0; i < BoardRegion.Width; i++)
            {
                var offset = CellSize * i + CellSize;
                var columnHeader = CreateHeaderCell(offset, 0, (i + 1).ToString());
                var rowHeader = CreateHeaderCell(0, offset, (i + 1).ToString());

                Controls.Add(columnHeader);
                Controls.Add(rowHeader);
            }

        }

        /// <summary>
        ///     Creates a board layout
        /// </summary>
        private void CreateBoard()
        {
            SuspendLayout();

            var boardSize = new Size(CellSize * BoardRegion.Width + CellSize, CellSize * BoardRegion.Height + CellSize);
            base.MinimumSize = boardSize;
            base.MaximumSize = boardSize;

            CreateHeaders();

            var points = BoardRegion.GetPoints();

            foreach(var point in points)
            {
                var cell = new BoardCell(point.X, point.Y)
                               {
                                   Top = point.X * CellSize + CellSize,
                                   Left = point.Y * CellSize + CellSize,
                                   Width = CellSize,
                                   Height = CellSize,
                                   State = BoardCellState.Normal
                               };
                _cells[point.X, point.Y] = cell;
                cell.MouseDown += OnCellMouseDown;
                cell.DragEnter += OnCellDragEnter;
                cell.DragLeave += OnCellDragLeave;
                cell.DragDrop += OnCellDragDrop;
                cell.QueryContinueDrag += OnCellQueryContinueDrag;
                cell.Click += OnCellClick;
                Controls.Add(cell);
            }

            ResumeLayout();
        }

        private void OnCellClick(object sender, EventArgs e)
        {
            if (Mode != BoardMode.Game)
                return;

            var handler = OnClick;
            if (handler == null)
                return;

            var cell = (BoardCell)sender;
            var eventArgs = new BoardCellClickEventErgs(cell.X, cell.Y);
            handler(this, eventArgs);
        }

        /// <summary>
        ///     Get a ship a a given location
        /// </summary>
        /// <param name="x">X coordinate to check ship at</param>
        /// <param name="y">y coordinate to check ship at</param>
        /// <returns><see cref="Ship"/></returns>
        private Ship GetShipAt(int x, int y)
        {
            return _ships.FirstOrDefault(ship => ship.IsLocatedAt(x, y));
        }

        /// <summary>
        ///     Handles <see cref="BoardCell"/>'s MouseDown event and initiates ship drag'n'drop    
        /// </summary>
        private void OnCellMouseDown(object sender, MouseEventArgs e)
        {
            if (Mode == BoardMode.Game || !_drawShips)
                return;

            var cell = (BoardCell)sender;
            var ship = GetShipAt(cell.X, cell.Y);

            if (ship == null)
            {
                return;
            }
            _draggedShip = DraggableShip.From(ship);
            cell.DoDragDrop(ship, DragDropEffects.Copy | DragDropEffects.Move);
        }

        /// <summary>
        ///     Gives feedback for ship rotation while dragging it
        /// </summary>
        private void OnCellQueryContinueDrag(object sender, QueryContinueDragEventArgs e)
        {
            // check Ctrl key state
            var shouldRotate = ((e.KeyState & 8) == 8);
            var isRotated = _draggedShip.IsOrientationModified;

            if ((shouldRotate && isRotated) || (!shouldRotate && !isRotated))
                return;

            var rect = _draggedShip.GetShipRegion();
            RedrawRegion(rect);

            _draggedShip.Rotate();
            _draggedShip.IsOrientationModified = !isRotated;

            var state = CanPlaceShip(_draggedShip, _draggedShip.X, _draggedShip.Y) ? BoardCellState.ShipDrag : BoardCellState.ShipDragInvalid;
            DrawShip(_draggedShip, state);
        }

        /// <summary>
        ///     Handles <see cref="BoardCell"/>'s DragEnter event
        /// </summary>
        private void OnCellDragEnter(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(typeof(Ship)))
            {
                var cell = (BoardCell)sender;
                _draggedShip.MoveTo(cell.X, cell.Y);

                var canPlaceShip = CanPlaceShip(_draggedShip, cell.X, cell.Y);
                var state = canPlaceShip ? BoardCellState.ShipDrag : BoardCellState.ShipDragInvalid;

                DrawShip(_draggedShip, state);

                e.Effect = canPlaceShip ? DragDropEffects.Move : DragDropEffects.None;
            }
            else
            {
                e.Effect = DragDropEffects.None;
            }
        }

        /// <summary>
        ///     Handles <see cref="BoardCell"/>'s DragLeave event
        /// </summary>
        private void OnCellDragLeave(object sender, EventArgs e)
        {
            var rect = _draggedShip.GetShipRegion();
            RedrawRegion(rect);

        }

        /// <summary>
        ///     Handles <see cref="BoardCell"/>'s DragDrop event
        /// </summary>
        private void OnCellDragDrop(object sender, DragEventArgs e)
        {
            var cell = (BoardCell)sender;
            if (e.Data.GetDataPresent(typeof(Ship)))
            {
                if (!CanPlaceShip(_draggedShip, cell.X, cell.Y))
                    return;

                var ship = _draggedShip.Source;
                _ships.Remove(ship);

                var rect = ship.GetShipRegion();
                RedrawRegion(rect);

                ship.Orientation = _draggedShip.Orientation;

                AddShip(ship, cell.X, cell.Y);
                _draggedShip = null;

            }
            else
            {
                e.Effect = DragDropEffects.None;
            }
        }

        /// <summary>
        ///     Adds a ship on a board
        /// </summary>
        /// <param name="ship">Ship to add</param>
        /// <param name="x">X coordinate of where to add ship</param>
        /// <param name="y">Y coordinate of where to add ship</param>
        public void AddShip(Ship ship, int x, int y)
        {
            ship.MoveTo(x, y);

            _ships.Add(ship);
            DrawShip(ship, BoardCellState.Ship);
        }

        /// <summary>
        ///     Returns true if given shio van be placed at a given location
        /// </summary>
        /// <param name="ship">Ship to check</param>
        /// <param name="x">X coordinate to check</param>
        /// <param name="y">Y coordinate to check</param>
        private bool CanPlaceShip(Ship ship, int x, int y)
        {
            var shipRegion = ship.GetShipRegion();

            shipRegion.MoveTo(x, y);

            if (!BoardRegion.Contains(shipRegion))
                return false;

            shipRegion.Inflate(1, 1);

            foreach (var s in _ships)
            {
                if (ship is DraggableShip && s == ((DraggableShip)ship).Source)
                {
                    continue;
                }

                if (s.GetShipRegion().IntersectsWith(shipRegion))
                {
                    return false;
                }
            }


            return true;
        }

        /// <summary>
        ///     Redraws a given region on a board
        /// </summary>
        /// <param name="region">A region to redraw</param>
        private void RedrawRegion(Rect region)
        {
            SuspendLayout();

            var points = region.GetPoints();
            foreach (var point in points)
            {
                if (!BoardRegion.Contains(point))
                {
                    continue;
                }

                var ship = GetShipAt(point.X, point.Y);
                _cells[point.X, point.Y].State = ship == null ? BoardCellState.Normal : BoardCellState.Ship;
            }

            ResumeLayout();
        }


        public void ShowShips()
        {
            foreach(var ship in _ships)
            {
                var shipPoints = ship.GetShipRegion().GetPoints();

                foreach(var point in shipPoints)
                {
                    var cell = _cells[point.X, point.Y];
                    if (cell.State != BoardCellState.Normal)
                        continue;

                    cell.State = BoardCellState.Ship;                        
                }
            }
        }

        /// <summary>
        ///     Draws a ship on a board
        /// </summary>
        /// <param name="ship">Ship to draw</param>
        /// <param name="state">Ship state to draw</param>
        private void DrawShip(Ship ship, BoardCellState state)
        {
            DrawShip(ship, state, false);
        }

        /// <summary>
        ///     Draws a ship on a board
        /// </summary>
        /// <param name="ship">Ship to draw</param>
        /// <param name="state">Ship state to draw</param>
        /// <param name="force">True to force ship drawing</param>
        private void DrawShip(Ship ship, BoardCellState state, bool force)
        {
            if (!_drawShips && !force)
                return;

            var points = ship.GetShipRegion().GetPoints();

            foreach (var point in points)
            {
                if (BoardRegion.Contains(point))
                {
                    _cells[point.X, point.Y].State = state;
                }
            }
        }

        /// <summary>
        ///     Removes all ships from board
        /// </summary>
        public void ClearBoard()
        {
            SuspendLayout();

            _ships.Clear();

            var points = BoardRegion.GetPoints();
            foreach (var point in points)
            {
                _cells[point.X, point.Y].State = BoardCellState.Normal;
            }

            ResumeLayout();
        }

        /// <summary>
        ///     Adds random ships on a board
        /// </summary>
        public void AddRandomShips()
        {
            SuspendLayout();

            ClearBoard();

            var ships = GetNewShips();

            foreach (var ship in ships)
            {
                var shipAdded = false;

                while (!shipAdded)
                {
                    var x = _rnd.Next(10);
                    var y = _rnd.Next(10);

                    if (!CanPlaceShip(ship, x, y))
                        continue;
                    
                    AddShip(ship, x, y);
                    shipAdded = true;
                }
            }

            ResumeLayout();
        }

        /// <summary>
        ///     Returns a list of ships with random locations. 
        ///     This is usualy used to begin new game     
        /// </summary>
        private IList<Ship> GetNewShips()
        {
            var ships = new List<Ship>
                        {
                            new Ship(4){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(3){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(3){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(2){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(2){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(2){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(1){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(1){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(1){Orientation = (ShipOrientation)_rnd.Next(2)},
                            new Ship(1){Orientation = (ShipOrientation)_rnd.Next(2)}
                        };

            return ships;
        }

        public BoardMode Mode { get; set; }


        public ShotResult OpenentShotAt(int x, int y)
        {
            var ship = GetShipAt(x, y);

            if (ship == null)
            {
                _cells[x, y].State = BoardCellState.MissedShot;
                return ShotResult.Missed;
            }
            _cells[x, y].State = BoardCellState.ShotShip;

            ship.HitCount++;

            if (ship.IsDrowned)
                DrawShip(ship, BoardCellState.ShowDrowned, true);

            return ship.IsDrowned ? ShotResult.ShipDrowned : ShotResult.ShipHit;
        }

        public new event EventHandler<BoardCellClickEventErgs> OnClick;

        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);
            Font = Parent.Font;
            Debug.WriteLine(Font.Name);
        }

    }
}

File: SeaBattle/SeaBattle.Cpp/Board.cpp
#include "stdafx.h"
#include "Board.h"
#include "Rect.h"

Board::Board()
{
   _drawShips = true;
	_cells = gcnew array<BoardCell^, 2>(10, 10);
	_ships = gcnew List<Ship^>;
	_rnd = gcnew Random(DateTime::Now.Millisecond);
	Mode = BoardMode::Design;
	Margin = Padding.Empty;

	CreateBoard();
}

Board::Board(bool drawShips)
{
	_drawShips = drawShips;
	_cells = gcnew array<BoardCell^, 2>(10, 10);
	_ships = gcnew List<Ship^>;
	_rnd = gcnew Random(DateTime::Now.Millisecond);
	Mode = BoardMode::Design;
	Margin = Padding.Empty;

	CreateBoard();
}

Label^ Board::CreateHeaderCell(int x, int y, String^ text)
{
	Label^ cell = gcnew Label;
    
    cell->AutoSize = false;
    cell->BackColor = Color::Transparent;
    cell->TextAlign = ContentAlignment::MiddleCenter;
    cell->Text = text;
    cell->Location = Point(x, y);
    cell->Width = CellSize;
    cell->Height = CellSize;

	return cell;
};

void Board::CreateHeaders()
{
	for (int i = 0; i < BoardRegion->Width; i++)
	{
		int offset = CellSize * i + CellSize;
		Label^ columnHeader = CreateHeaderCell(offset, 0, (i + 1).ToString());
		Label^ rowHeader = CreateHeaderCell(0, offset, (i + 1).ToString());

		Controls->Add(columnHeader);
		Controls->Add(rowHeader);
	}
};

void Board::CreateBoard()
{
	SuspendLayout();

    CreateHeaders();

	IList<Point>^ points = BoardRegion->GetPoints();

	for each (Point point in points)
	{
		BoardCell^ cell = gcnew BoardCell(point.X, point.Y);
        cell->Top = point.X * CellSize + CellSize;
        cell->Left = point.Y * CellSize + CellSize;
        cell->Width = CellSize;
        cell->Height = CellSize;
        cell->State = BoardCellState::Normal;
		
		cell->MouseDown += gcnew MouseEventHandler(this, &Board::OnCellMouseDown);
		cell->DragEnter += gcnew DragEventHandler(this, &Board::OnCellDragEnter);
		cell->DragLeave += gcnew EventHandler(this, &Board::OnCellDragLeave);
		cell->DragDrop += gcnew DragEventHandler(this, &Board::OnCellDragDrop);
		cell->QueryContinueDrag += gcnew QueryContinueDragEventHandler(this, &Board::OnCellQueryContinueDrag);
		cell->Click += gcnew EventHandler(this, &Board::OnCellClick);

		_cells[point.X, point.Y] = cell;
        
        Controls->Add(cell);
	}

    delete points;

    Drawing::Size boardSize = Drawing::Size(CellSize * BoardRegion->Width + CellSize, CellSize * BoardRegion->Height + CellSize);
	MinimumSize = boardSize;
	MaximumSize = boardSize;

	ResumeLayout();
};

void Board::RedrawRegion(Rect^ region)
{
	SuspendLayout();

	IList<Point>^ points = region->GetPoints();
	for each (Point point in points)
	{
		if (!BoardRegion->Contains(point))
		{
			continue;
		}

		Ship^ ship = GetShipAt(point.X, point.Y);
		_cells[point.X, point.Y]->State = ship == nullptr ? BoardCellState::Normal : BoardCellState::Ship;
	}

    delete points;

	ResumeLayout();
};

void Board::DrawShip(Ship^ ship, BoardCellState state)
{
	DrawShip(ship, state, false);
};

void Board::DrawShip(Ship^ ship, BoardCellState state, bool force)
{
	if (!_drawShips && !force)
		return;

	IList<Point>^ points = ship->GetShipRegion()->GetPoints();

	for each (Point point in points)
	{
        System::Diagnostics::Debug::WriteLine("B");
		if (BoardRegion->Contains(point))
		{
			_cells[point.X, point.Y]->State = state;
		}
	}

    delete points;
};

void Board::OnCellClick(Object^ sender, EventArgs^ e)
{
	if (Mode != BoardMode::Game)
		return;

    BoardCell^ cell = safe_cast<BoardCell^>(sender);
	BoardCellClickEventArgs^ eventArgs = gcnew BoardCellClickEventArgs(cell->X, cell->Y);
	OnClick(this, eventArgs);
};

void Board::OnCellMouseDown(Object^ sender, MouseEventArgs^ e)
{
	if (Mode == BoardMode::Game || !_drawShips)
		return;

	BoardCell^ cell = safe_cast<BoardCell^>(sender);
	Ship^ ship = GetShipAt(cell->X, cell->Y);

	if (ship == nullptr)
	{
		return;
	}
	_draggedShip = DraggableShip::From(ship);
	cell->DoDragDrop(ship, DragDropEffects::Copy | DragDropEffects::Move);
};

void Board::OnCellQueryContinueDrag(Object^ sender, QueryContinueDragEventArgs^ e)
{
	// check Ctrl key state
	bool shouldRotate = ((e->KeyState & 8) == 8);
	bool isRotated = _draggedShip->IsOrientationModified;

	if ((shouldRotate && isRotated) || (!shouldRotate && !isRotated))
		return;

	Rect^ rect = _draggedShip->GetShipRegion();
	RedrawRegion(rect);

    delete rect;

	_draggedShip->Rotate();
	_draggedShip->IsOrientationModified = !isRotated;

	BoardCellState state = CanPlaceShip(_draggedShip, _draggedShip->X, _draggedShip->Y) ? BoardCellState::ShipDrag : BoardCellState::ShipDragInvalid;
	DrawShip(_draggedShip, state);
};

void Board::OnCellDragEnter(Object^ sender, DragEventArgs^ e)
{
   
	if (e->Data->GetDataPresent(Ship::typeid))
	{
		BoardCell^ cell = safe_cast<BoardCell^>(sender);
		_draggedShip->MoveTo(cell->X, cell->Y);

		bool canPlaceShip = CanPlaceShip(_draggedShip, cell->X, cell->Y);
		BoardCellState state = canPlaceShip ? BoardCellState::ShipDrag : BoardCellState::ShipDragInvalid;

		DrawShip(_draggedShip, state);

		e->Effect = canPlaceShip ? DragDropEffects::Move : DragDropEffects::None;
	}
	else
	{
		e->Effect = DragDropEffects::None;
	}
};

void Board::OnCellDragLeave(Object^ sender, EventArgs^ e)
{
	Rect^ rect = _draggedShip->GetShipRegion();
	RedrawRegion(rect);
    delete rect;
};

void Board::OnCellDragDrop(Object^ sender, DragEventArgs^ e)
{
   BoardCell^ cell = safe_cast<BoardCell^>(sender);
   if (e->Data->GetDataPresent(Ship::typeid))
   {
	   if (!CanPlaceShip(_draggedShip, cell->X, cell->Y))
		   return;

	   Ship^ ship = _draggedShip->Source;
	   _ships->Remove(ship);

	   Rect^ rect = ship->GetShipRegion();
	   RedrawRegion(rect);
       delete rect;

	   ship->Orientation = _draggedShip->Orientation;

	   AddShip(ship, cell->X, cell->Y);
	   delete _draggedShip;
   }
   else
   {
	   e->Effect = DragDropEffects::None;
   }
};

Ship^ Board::GetShipAt(int x, int y)
{
    for each(Ship^ ship in _ships)
    {
        if (ship->IsLocatedAt(x, y))
        {
            return ship;
        }
    }

    return nullptr;
};

bool Board::CanPlaceShip(Ship^ ship, int x, int y)
{
	Rect^ shipRegion = ship->GetShipRegion();

	shipRegion->MoveTo(x, y);

	if (!BoardRegion->Contains(shipRegion))
    {
        delete shipRegion;
        return false;
    }
	shipRegion->Inflate(1, 1);

	for each (Ship^ s in _ships)
	{
		if (dynamic_cast<DraggableShip^>(ship) != nullptr && s == (safe_cast<DraggableShip^>(ship))->Source)
		{
            delete shipRegion;
			continue;
		}

		if (s->GetShipRegion()->IntersectsWith(shipRegion))
		{
            delete shipRegion;
			return false;
		}
	}

    delete shipRegion;
	return true;
};

IList<Ship^>^ Board::GetNewShips()
{
	List<Ship^>^ ships = gcnew List<Ship^>;
    ships->Add(gcnew Ship(4, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(3, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(3, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(2, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(2, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(2, safe_cast<ShipOrientation>(_rnd->Next(2))));
    ships->Add(gcnew Ship(1, ShipOrientation::Horizontal));
    ships->Add(gcnew Ship(1, ShipOrientation::Horizontal));
    ships->Add(gcnew Ship(1, ShipOrientation::Horizontal));
    ships->Add(gcnew Ship(1, ShipOrientation::Horizontal));

	return ships;
};

void Board::ClearBoard()
{
	SuspendLayout();

	_ships->Clear();

	IList<Point>^ points = BoardRegion->GetPoints();
	for each (Point point in points)
	{
		_cells[point.X, point.Y]->State = BoardCellState::Normal;
	}

    delete points;

	ResumeLayout();
};

void Board::AddShip(Ship^ ship, int x, int y)
{
	ship->MoveTo(x, y);

	_ships->Add(ship);
	DrawShip(ship, BoardCellState::Ship);
};

void Board::ShowShips()
{
	for each (Ship^ ship in _ships)
	{
		IList<Point>^ shipPoints = ship->GetShipRegion()->GetPoints();

		for each (Point point in shipPoints)
		{
			BoardCell^ cell = _cells[point.X, point.Y];
			if (cell->State != BoardCellState::Normal)
				continue;

			cell->State = BoardCellState::Ship;
		}

        delete shipPoints;
	}
};

void Board::AddRandomShips()
{
	SuspendLayout();

	ClearBoard();

	IList<Ship^>^ ships = GetNewShips();

	for each (Ship^ ship in ships)
	{
		bool shipAdded = false;

		while (!shipAdded)
		{
			int x = _rnd->Next(10);
			int y = _rnd->Next(10);

			if (!CanPlaceShip(ship, x, y))
				continue;

			AddShip(ship, x, y);
			shipAdded = true;
		}
	}

	ResumeLayout();
};

ShotResult Board::OpenentShotAt(int x, int y)
{
	Ship^ ship = GetShipAt(x, y);

	if (ship == nullptr)
	{
		_cells[x, y]->State = BoardCellState::MissedShot;
		return ShotResult::Missed;
	}
	_cells[x, y]->State = BoardCellState::ShotShip;

	ship->HitCount++;

	if (ship->IsDrowned)
		DrawShip(ship, BoardCellState::ShowDrowned, true);

	return ship->IsDrowned ? ShotResult::ShipDrowned : ShotResult::ShipHit;
}

void Board::OnParentChanged(EventArgs^ e)
{
	Control::OnParentChanged(e);
	Font = Parent->Font;
}






















